File: test/system/cp.mjs
// /bin/cp: the copy proc/io.h holds, which /bin/mv falls back to wherever the
// store will not rename.
// Part of the system suite; test/run.mjs runs the cases in order and
// doc/Testing.md has the rules they run by.

import { shows } from "./harness.mjs";

const { at, is, line } = shows(13900);

export function check() {
    // Relative names throughout: the screen is 60 columns and a command that
    // wraps puts its own tail in the output.
    line("mkdir /home/cp");
    line("cd /home/cp");
    line("mkdir d");
    line("echo one > a");
    line("echo two > d/b");

    // A file to a new name, and into a directory that exists.
    is("cp a c; cat c", "one");
    is("cp a d; cat d/a", "one");

    // The source is still there — the whole difference from mv.
    is("cat a", "one");

    // A directory needs -r, and says so rather than copying nothing.
    is("cp d e", "cp: d: is a directory");
    is("cp -r d e; cat e/b", "two");

    // -n declines an existing destination and still succeeds; without it the
    // destination is taken.
    line("echo new > n");
    is("cp -n n c; cat c", "one");
    is("cp n c; cat c", "new");

    // A link is recreated rather than followed: copy_tree descends on
    // SYS_KIND_DIR alone, so no cycle guard is needed.
    line("ln -s /home/cp/a d/l");
    line("cp -r d f");
    is("test -h f/l && cat f/l", "one");

    // A destination that already holds the copy is merged into rather than
    // removed and remade: a name only it has survives. The link is replaced
    // rather than refused, so a second copy is not an error either.
    line("mkdir m");
    line("cp -r d m");
    line("echo keep > m/d/keep");
    is("cp -r d m; cat m/d/keep m/d/b", "keep|two");
    is("test -h m/d/l && cat m/d/l", "one");

    // A file where a directory must go is told apart from a directory that is
    // already there: the stat make_dir's Err(Exists) does not give.
    line("mkdir d/sub");
    line("echo x > m/d/sub");
    is("cp -r d m", "cp: d: already exists");

    // And a directory where a file must go, which is the open's answer.
    line("rm m/d/sub m/d/b");
    line("mkdir m/d/b");
    is("cp -r d m", "cp: d: is a directory");
    line("rm -r m/d/b d/sub");

    // -n and -i still decide at the named destination, not per file inside it.
    is("cp -rn d m; cat m/d/keep", "keep");
    is("echo n | cp -ri d m", "overwrite /home/cp/m/d?");

    // -i on a file: the answer decides, and -f overrides the ask, as in mv.
    // Only the refusal above had ever run, though cp.cpp asks on the same
    // `!force && ask` mv does — doc/TODO.md D3 is why that is worth pinning.
    line("echo old > y1");
    line("echo fresh > y2");
    is("echo n | cp -i y2 y1", "overwrite /home/cp/y1?");
    is("cat y1", "old");
    is("echo y | cp -i y2 y1", "overwrite /home/cp/y1?");
    is("cat y1", "fresh");
    line("echo old > y1");
    is("echo n | cp -fi y2 y1", "");
    is("cat y1", "fresh");
    line("rm y1 y2");

    // Bare is asking rather than getting it wrong: 0, and nothing removed.
    is("rm; echo $?", "Usage:|    rm [-r] <path>...|Options:|" +
                      "    -r    remove directories, and what is in them|0");
    is("cat a", "one");

    // Several sources need a directory to land in.
    is("cp a n c", "Usage:|    cp [-r] [-fi] [-n] <src> <dst>|    cp [-r] [-fi] [-n] <src>... <dir>|Options:|" +
                "    -r    copy directories, and what is in them|" +
                "    -f    replace what stands at the destination|" +
                "    -i    ask before replacing it|" +
                "    -n    keep it, and copy nothing over it");
    line("mkdir g");
    line("cp a n g");
    is("cat g/a g/n", "one|new");

    // Onto itself, and into itself: refused rather than looping.
    is("cp a a", "cp: a: invalid");
    is("cp -r d d/in", "cp: d: cannot copy a directory into itself");

    // A source that is not there is reported, and the status says so.
    is("cp nope z; echo $?", "cp: nope: not found|1");

    line("cd /home"); // the session is cumulative: leave the cwd as it was
}

File: src/cmd/cp.cpp
#include "fs/path.h"
#include "proc/io.h"
#include "proc/opt.h"
#include "proc/usage.h"

// The copy itself is proc/io.h's, shared with /bin/mv, which falls back to it
// wherever the store will not rename.

namespace {

constexpr Str USAGE =
    "Usage:\n"
    "    cp [-r] [-fi] [-n] <src> <dst>\n"
    "    cp [-r] [-fi] [-n] <src>... <dir>\n"
    "Options:\n"
    "    -r    copy directories, and what is in them\n"
    "    -f    replace what stands at the destination\n"
    "    -i    ask before replacing it\n"
    "    -n    keep it, and copy nothing over it\n";

struct Flags {
    bool recurse = false; // -r
    bool force   = false; // -f
    bool ask     = false; // -i
    bool no_clob = false; // -n
};

// "overwrite <dst>? " and one line of the answer. mv.cpp's rule: anything but a
// leading y declines.
Task<Result<bool>> confirm(Str dst, LineReader &answers)
{
    co_await write_all(SYS_STDERR, "overwrite ");
    co_await write_all(SYS_STDERR, dst);
    co_await write_all(SYS_STDERR, "? ");

    String line;
    Result<bool> r = co_await answers.next(line);
    if (r.is_err())
        co_return Err(r.error());
    co_return r.value() && !line.empty() && (line.str()[0] == 'y' || line.str()[0] == 'Y');
}

// One source to one destination, both absolute.
Task<Result<void>> copy_one(Str from, Str to, Flags f, LineReader &answers)
{
    if (from == to)
        co_return Err(Error::Invalid);
    if (path_under(from, to))
        co_return Err(Error::Invalid);

    Result<FileInfo> src = Err(Error::NoMemory);
    if (Task<Result<FileInfo>> t = stat_of(from, false))
        src = co_await t;
    if (src.is_err())
        co_return Err(src.error());

    if (src.value().kind == SYS_KIND_DIR && !f.recurse)
        co_return Err(Error::IsDir);

    Result<FileInfo> dst = Err(Error::NoMemory);
    if (Task<Result<FileInfo>> t = stat_of(to, false))
        dst = co_await t;
    if (dst.is_err() && dst.error() != Error::NotFound)
        co_return Err(dst.error());

    if (dst.is_ok()) {
        if (f.no_clob)
            co_return {};
        if (!f.force && f.ask) {
            Result<bool> yes = Err(Error::NoMemory);
            if (Task<Result<bool>> t = confirm(to, answers))
                yes = co_await t;
            if (yes.is_err())
                co_return Err(yes.error());
            if (!yes.value())
                co_return {};
        }
        // Two directories merge; every other pair needs the name cleared, and
        // a link is replaced rather than written through.
        bool merge = src.value().kind == SYS_KIND_DIR && dst.value().kind == SYS_KIND_DIR;
        if (!merge && (src.value().kind != SYS_KIND_FILE || dst.value().kind != SYS_KIND_FILE)) {
            Result<void> d = Err(Error::NoMemory);
            if (Task<Result<void>> t = remove_path(to, true))
                d = co_await t;
            if (d.is_err())
                co_return d;
        }
    }

    if (src.value().kind == SYS_KIND_LINK) {
        Result<String> target = Err(Error::NoMemory);
        if (Task<Result<String>> t = read_link(from))
            target = co_await t;
        if (target.is_err())
            co_return Err(target.error());
        if (Task<Result<void>> t = make_link(target.value().str(), to))
            co_return co_await t;
        co_return Err(Error::NoMemory);
    }
    if (src.value().kind == SYS_KIND_DIR) {
        if (Task<Result<void>> t = copy_tree(from, to))
            co_return co_await t;
        co_return Err(Error::NoMemory);
    }
    if (Task<Result<void>> t = copy_file(from, to))
        co_return co_await t;
    co_return Err(Error::NoMemory);
}

} // namespace

Task<i32> proc_main(Args args)
{
    if (args.size() == 1 || help_asked(args))
        co_return co_await usage_asked(USAGE);

    Flags f;
    OptParse p(args, Opts{ "rRfin", "" });
    for (Opt o;;) {
        Result<bool> r = p.next(o);
        if (r.is_err()) {
            co_await write_all(SYS_STDERR, "cp: bad option\n");
            co_return co_await usage_error(USAGE);
        }
        if (!r.value())
            break;
        if (o.name == 'r' || o.name == 'R')
            f.recurse = true;
        else if (o.name == 'f')
            f.force = true;
        else if (o.name == 'i')
            f.ask = true;
        else
            f.no_clob = true;
    }

    Args rest = p.rest();
    if (rest.size() < 2)
        co_return co_await usage_error(USAGE);

    // Absolute throughout: copy_one compares its two paths, and a relative one
    // would not compare.
    Result<String> here = Err(Error::NoMemory);
    if (Task<Result<String>> t = cwd_get())
        here = co_await t;
    if (here.is_err())
        co_return here.error() == Error::Cancelled ? 130 : 1;

    String dest;
    if (path_resolve(here.value().str(), rest[rest.size() - 1], dest).is_err())
        co_return 1;

    // With one source the last operand is the new name; with more it is the
    // directory they go in.
    Result<FileInfo> s = Err(Error::NoMemory);
    if (Task<Result<FileInfo>> t = stat_of(dest.str(), true))
        s = co_await t;
    if (s.is_err() && s.error() == Error::Cancelled)
        co_return 130;

    bool into_dir = s.is_ok() && s.value().kind == SYS_KIND_DIR;
    if (!into_dir && rest.size() > 2) {
        co_return co_await usage_error(USAGE);
    }

    Input replies(Args{}, SYS_STDIN, "cp");
    LineReader answers(replies);

    i32 status = 0;
    for (usize i = 0; i + 1 < rest.size(); i++) {
        String from, to;
        if (path_resolve(here.value().str(), rest[i], from).is_err())
            co_return 1;
        if (into_dir) {
            if (path_join(dest.str(), path_basename(from.str()), to).is_err())
                co_return 1;
        } else if (!to.assign(dest.str())) {
            co_return 1;
        }

        if (from.str() != to.str() && path_under(from.str(), to.str())) {
            co_await write_all(SYS_STDERR, "cp: ");
            co_await write_all(SYS_STDERR, rest[i]);
            co_await write_all(SYS_STDERR, ": cannot copy a directory into itself\n");
            status = 1;
            continue;
        }

        Result<void> r = Err(Error::NoMemory);
        if (Task<Result<void>> t = copy_one(from.str(), to.str(), f, answers))
            r = co_await t;
        if (r.is_ok())
            continue;
        if (r.error() == Error::Cancelled)
            co_return 130;

        status = 1;
        if (Task<void> e = errln("cp", rest[i], r.error()))
            co_await e;
    }
    co_return status;
}
